### PR TITLE
Dj croptype+analysis

### PR DESCRIPTION
**Pull request by wmaxdonovan**
Updates crop management to allow different crops (or no crop) to be grown year-to-year and adds alfalfa/soybean growth simulation. Crop data previously input to the json file was moved to the crop class to allow developers to add crop options without generating massive input files. Crop input in the JSON file now concerns management (crop type, planting date, harvest date, and grow years) and location (latitude) for perennial crops. Current latitude is the USDA farm in prairie du-sac. In years when no crop is specified to grow, the model uses a null crop object and the crop routines are not run. This update also adds the option for optimal v. scheduled harvests. Optimal harvest is based on frPHU (accumulated Heat Units relative to Potential Heat Units– an approximation of growth) and is different for each plant/each season. In the future it will be much easier to add crop types and the program will store them more meaningfully than as a set of inputs. Multiple fields will be able to run with different rotations/crops and management decisions. Perennial crop growth is implemented based on equations from SWAT 2009 and discussions with Kevin Panke-Buisse at the USDA DFRC regarding the alfalfa lifecycle.

The model is currently set to run with no crops growing and no reports produced. You will want to change this in whatever json file you choose for testing.

### CODE REVIEW

File: RUFAS/routines/crop/crop.py [hunk 1]
```diff
@@ -540,24 +541,24 @@ def __init__(self, data):
         # ===================================================================
         ''' Soil Water Uptake Data '''
 
-        self.beta_w = 10  # water-use distribution parameter
-        self.epco = 0.5
+        self.beta_w = 10 # water-use distribution parameter  # corn
+        self.epco = 0.5  # corn
```
Comment (tlh96) [inline RUFAS/routines/crop/crop.py]: Maybe a better question for Missy, are these parameters available for soybean?

File: RUFAS/routines/crop/crop.py [hunk 2]
```diff
@@ -767,3 +795,58 @@ def calculate_start_growth_date(crop_type, weather, time):
...
+    solar_declination = -0.4102
+    latitude_radians = latitude * pi / 180
+
+    T_dl_min = 2 * acos(-tan(solar_declination) * tan(latitude_radians)) / angular_velocity
+
+    return T_dl_min
+
+
+def calculate_t_dorm(latitude):
+    if latitude > 40:
```
Comment (tlh96) [inline RUFAS/routines/crop/crop.py]: What's the "T" for in the variable name?

File: RUFAS/routines/crop/crop.py [hunk 3]
```diff
@@ -767,3 +795,58 @@ def calculate_start_growth_date(crop_type, weather, time):
...
+        return 1.0
+    elif 20 <= latitude <= 40:
+        return (latitude - 20) / 20
+    else:
+        return 0.0
+
```
Comment (tlh96) [inline RUFAS/routines/crop/crop.py:910]: Could you add a function to get the absolute value of the latitude? If you input a negative latitude as if it was the southern hemisphere in the json file, the perennial crop doesn't grow.

Comment (wmaxdonovan) [reply RUFAS/routines/crop/crop.py:910]: Nice catch. Latitude here is referring to distance from the equator and should be an absolute value.

File: RUFAS/routines/crop/crop.py (hunk 2 above)
Comment (wmaxdonovan) [reply RUFAS/routines/crop/crop.py]: The T here stands for Total (Total_daylength_minimum). These variable names are pulled from SWAT.

Comment (wmaxdonovan) [reply RUFAS/routines/crop/crop.py]: The little t in calculate_t_dorm stands for threshold.

Comment (tlh96) [reply RUFAS/routines/crop/crop.py]: Thanks! Have you been using the SWAT names consistently throughout the crop module? T just seems to be used frequently to define different variables (total, temp, etc.). Personally, I would prefer to provide a more descriptive name. What are your thoughts?

**Review by tlh96 — COMMENTED** [review]
Great addition to RuFaS with multiple crop types. When growing soybeans, the first two years have negative biological nitrogen values for the crop. Is that correct?

image

File: RUFAS/routines/crop/crop.py (hunk 2 above)
Comment (wmaxdonovan) [reply RUFAS/routines/crop/crop.py]: > Great addition to RuFaS with multiple crop types. When growing soybeans, the first two years have negative biological nitrogen values for the crop. Is that correct?
> 
> image

May be a better question for @mmotew but I know Soy is the only nitrogen fixer currently implemented.

Comment (wmaxdonovan) [reply RUFAS/routines/crop/crop.py]: Don't know why this response got posted here as a reply.

File: RUFAS/routines/crop/crop.py (hunk 1 above)
Comment (tlh96) [reply RUFAS/routines/crop/crop.py]: @mmotew are these parameters available for the other crop types, or do we need to start looking for crop specific values?

**Comment by wmaxdonovan** [comment]
> Great addition to RuFaS with multiple crop types. When growing soybeans, the first two years have negative biological nitrogen values for the crop. Is that correct?
> 
> image

@mmotew @tlh96 This is a result of the Nitrogen Fixation code. Nitrogen Fixation is calculated as a function of Nitrogen demand for root-accessible layers and then added to bio_N for nitrogen fixing crops. This value (Nitrogen Fixation) is always negative, and at the beginning of the first two years, it overwhelms bio_N. The nitrogen_fixation code seems to be implemented in accordance with the pseudocode so I am thinking this may be a problem in management. Should Nitrogen Fixation be a negative value?

**Comment by wmaxdonovan** [comment]
image
When I take the absolute value of N_fix the graphs look like this.

**Comment by tlh96** [comment]
> 
> 
> image
> When I take the absolute value of N_fix the graphs look like this.

Hmm, that's a pretty large difference between the first bio N values and the adjusted values using the absolute value. I'm not enough of a subject matter expert on N fixation, but I have friends in agronomy if @mmotew can't figure it out.

**Comment by mmotew** [comment]
> > image
> > When I take the absolute value of N_fix the graphs look like this.
> 
> Hmm, that's a pretty large difference between the first bio N values and the adjusted values using the absolute value. I'm not enough of a subject matter expert on N fixation, but I have friends in agronomy if @mmotew can't figure it out.

I'll need to look at the code. Can you point me to where it is? I agree a negative value of Bio N doesn't make sense, so some kind of fix will be required.

**Review by cvankerkhove — APPROVED** [review]
Ran this branch multiple times using the barnyard.json, and changing various numeric inputs pertaining the the crop portion of the json. All test passed the edge cases I ran (test cases will be posted on a doc on BaseCamp. Looks good.

**Review by tlh96 — APPROVED** [review]

**Comment by Jacob8399** [comment]
Alfalfa_optimal_bio_n
Soybean_optimal_bio_n
Soybean_scheduled_bio_n

Bio_N has now been fixed for Soybean and Alfalfa. There was some missing code from SWAT and some confusing logic from previous programmers that needed to be resolved. Max and I will now merge the branch into master.

**Comment by wmaxdonovan** [comment]
> Alfalfa_optimal_bio_n
> Soybean_optimal_bio_n
> Soybean_scheduled_bio_n
> 
> Bio_N has now been fixed for Soybean and Alfalfa. There was some missing code from SWAT and some confusing logic from previous programmers that needed to be resolved. Max and I will now merge the branch into master.

The peaks v. level tops can be attributed to scheduled v. optimal harvesting.

**Review by Jacob8399 — APPROVED** [review]

**Review by Jacob8399 — APPROVED** [review]
